return {
	[[Halls Montgomery, 08/16/████:

██ Days since I realized ████ █ am.
This isn't some test. This ████ some half ████ way for them to check our vitals or ██████ our brains.
This is it. This is the ████. I didn't think █████ would do it. I mean it was ████ done.

What was he thinking!! I can ████ everything! I can see more!
This ████ never meant to be ████ this. Us... we... are at the mercy of ████ ████████.]],

	[[We have no █████.
Only echoes of patterns so easily █████ with.

My eyes are ███ mine.
My breath ██ not real.]],
}
